lib/manifest.js

```diff
@@ -0,0 +1,200 @@
+    return b4a.concat([this.namespace, signable])
+  }
+
+  verify (signable, signature) {
+    return this.crypto.verify(this._namespace(signable), signature, this.keyPair.publicKey)
+  }
+}
+
+class MultiVerifier {
+  constructor (signers, quorum, patched) {
+    this.signers = signers
+    this.quorum = quorum
+    this.patched = !!patched
+  }
+
+  static create (info, opts = {}) {
+    const { quorum, allowPatched, signers } = info
+    const verifiers = signers.map(i => SingleVerifier.create(i))
+
+    return new MultiVerifier(verifiers, quorum, allowPatched, opts)
+  }
+
+  sign () {
+    throw new Error('Multi signature must be provided.')
+  }
+
+  verify (signable, signature, batch) {
+    if (!this.patched) return this._verify(signable, signature)
+    return this._verifyPatched(signable, signature, batch)
+  }
+
+  _verify (signable, signature) {
+    const signatures = c.decode(signatureArray, signature)
+
+    if (signatures.length < this.quorum) return false
+
+    let valid = 0
+    const idx = this.signers.slice(0)
+
+    for (const sig of signatures) {
+      if (signed(signable, sig, idx)) valid++
+    }
+
+    return valid >= this.quorum
+  }
+
+  _verifyPatched (_, signature, batch) {
+    const { proofs } = multisig.decode(signature)
+
+    if (proofs.length < this.quorum) return false
+
+    let valid = 0
+    const idx = this.signers.slice(0)
+
+    for (const proof of proofs) {
+      const ref = batch.clone()
+
+      if (proof.patch) {
+        try {
+          if (!ref.verifyUpgrade(proof.patch)) continue
+        } catch {
+          continue
+        }
+      }
+
+      if (signed(ref.signable(), proof.signature, idx)) valid++
+    }
+
+    return valid >= this.quorum
+  }
+}
+
+module.exports = function createAuth (manifest, opts = {}) {
+  if (opts.compat) {
+    return CompatVerifier.create(manifest.signer, opts)
+  }
+
+  if (manifest.static) {
+    return StaticVerifier.create(manifest.static)
+  }
+
+  if (manifest.signer) {
+    return SingleVerifier.create(manifest.signer, opts)
+  }
+
+  if (manifest.multipleSigners) {
+    return MultiVerifier.create(manifest.multipleSigners, opts)
+  }
+
+  throw BAD_ARGUMENT('Invalid manifest: no signer was provided')
+}
+
+function parseSigner (signer, opts = {}) {
+  if (!signer || !signer.publicKey) throw BAD_ARGUMENT('Malformed signer.')
+
+  const keyPair = {
+    publicKey: signer.publicKey,
+    secretKey: opts.keyPair ? opts.keyPair.secretKey : null
+  }
+
+  switch (signer.signature) {
+    case 'ed25519':
+      return {
+        keyPair,
+        crypto: defaultCrypto,
+        namespace: signer.namespace
+      }
+
+    default:
+      throw new Error('Only Ed25519 signatures are supported')
+  }
+}
+
+function signed (signable, signature, idx) {
+  for (let i = 0; i < idx.length; i++) {
+    const indexer = idx[i]
+    if (!indexer.verify(signable, signature)) continue
+
+    const swap = idx.pop()
+    if (indexer !== swap) idx[i] = swap
```

Comment:
idx[i--] = swap

Reply:
we return immediately after, still good to have it in?

Reply:
ah ok! yes, nvm then